crates/tx5/src/proto.rs

```diff
@@ -0,0 +1,580 @@
+    /// once we receive the authorization, forward the rest of
+    /// the message payload.
+    NeedPermit {
+        /// First, request a permit to send the payload.
+        permit_req: BackBuf,
+
+        /// Second, send the actual payload chunks.
+        msg_payload: Vec<BackBuf>,
+    },
+
+    /// This message fit in a single payload chunk. We do not need
+    /// to request a permit ahead of time, so just send the chunk.
+    OneMessage(BackBuf),
+}
+
+/// Encode some data into the Tx5 protocol.
+pub(crate) fn proto_encode(data: &[u8]) -> Result<ProtoEncodeResult> {
+    const MAX: usize = (16 * 1024) - 4;
+    let len = data.len();
+
+    if len > MAX_PAYLOAD as usize {
+        return Err(Error::id("PayloadSizeOverflow"));
+    }
+
+    if len <= MAX {
+        let (a, b, c, d) = ProtoHeader::CompleteMessage(len as u32).encode()?;
+
+        let mut go_buf = tx5_go_pion::GoBuf::new()?;
+        go_buf.reserve(len + 4)?;
+        go_buf.extend(&[a, b, c, d])?;
+        go_buf.extend(data)?;
+
+        Ok(ProtoEncodeResult::OneMessage(BackBuf::from_raw(go_buf)))
+    } else {
+        let (a, b, c, d) = ProtoHeader::PermitRequest(len as u32).encode()?;
+        let permit_req = BackBuf::from_slice([a, b, c, d])?;
+
+        let mut msg_payload = Vec::new();
+        let mut cur = 0;
+
+        while len - cur > 0 {
+            let amt = std::cmp::min((16 * 1024) - 4, len - cur);
+
+            let (a, b, c, d) =
+                ProtoHeader::MultipartMessage(amt as u32).encode()?;
+
+            let mut go_buf = tx5_go_pion::GoBuf::new()?;
+            go_buf.reserve(amt + 4)?;
+            go_buf.extend(&[a, b, c, d])?;
+            go_buf.extend(&data[cur..cur + amt])?;
+
+            msg_payload.push(BackBuf::from_raw(go_buf));
+
+            cur += amt;
+        }
+
+        Ok(ProtoEncodeResult::NeedPermit {
+            permit_req,
+            msg_payload,
+        })
+    }
+}
+
+/// Result of decoding an incoming message chunk.
+#[derive(Debug, PartialEq)]
+pub(crate) enum ProtoDecodeResult {
+    /// Nothing needs to happen at the moment... continue receiving chunks.
+    Idle,
+
+    /// Received incoming message.
+    Message(Vec<u8>),
+
+    /// The remote node is requesting a permit to send us chunks of data.
+    RemotePermitRequest(u32),
+
+    /// The remote node has granted us a permit to send them chunks of data.
+    RemotePermitGrant(u32),
+}
+
+#[derive(Clone, Copy, PartialEq)]
+enum DecodeState {
+    NeedVersion,
+    Ready,
+    /// The REMOTE requested a permit from US.
+    /// Totally different from when we make a permit request of the remote : )
+    RemoteAwaitingPermit(u32),
+    ReceiveChunked,
+}
+
+/// Tx5 protocol decoder.
+pub(crate) struct ProtoDecoder {
+    state: DecodeState,
+    want_size: usize,
+    incoming: Vec<u8>,
+    want_remote_grant: bool,
+    did_error: bool,
+    grant_permit: Option<tokio::sync::OwnedSemaphorePermit>,
+    grant_notify: Option<tokio::sync::oneshot::Sender<()>>,
```

Comment:
I highly recommend folding as much of this extra state as possible into the `DecodeState` enum. It looks like most or all of it is correlated: `want_size`, `incoming`, and `grant_permit` all seem only relevant to `RemoteAwaitingPermit` and could be fields on that variant. It's so helpful when working with state machines to keep them constrained as possible, for so many reasons.

The ideal would be that `ProtoDecoder` and `DecodeState` are the same thing.

Reply:
I understand the desire to do this, but I've found the reality in rust doesn't actually work. When you match on the state it captures a reference to self, and makes it almost impossible to modify that state (switch to a different enum variant) while you're matching on it.

The complexity bloomed, and I had to pull in things like [replace_with](https://crates.io/crates/replace_with).

Reply:
Yeah, I have encountered that too...quite the bummer!